### PR TITLE
chore: remove prune_modes from BlockWriter

### PR DESCRIPTION
**Pull request by Rjected**
This removes the `prune_modes` argument from the `BlockWriter` trait, because the argument leaks a node implementation detail into the storage abstraction. Instead, the it is included as a field in `ProviderFactory`, as well as `DatabaseProvider`, although it is only used in the RW version of `DatabaseProvider`.

When the blockchain tree is initialized, the prune modes are set, so that we achieve the same behavior as before when writing.

### CODE REVIEW

File: crates/storage/provider/src/providers/database/mod.rs [hunk 1]
```diff
@@ -46,6 +46,8 @@ pub struct ProviderFactory<DB> {
     chain_spec: Arc<ChainSpec>,
     /// Static File Provider
     static_file_provider: StaticFileProvider,
+    /// Optional pruning configuration
+    prune_modes: Option<PruneModes>,
```
Comment (mattsse) [inline crates/storage/provider/src/providers/database/mod.rs:50]: does this even have to be an Option?

all the fields of this types are options

Comment (Rjected) [reply crates/storage/provider/src/providers/database/mod.rs:50]: hmmm, I guess it doesn't need to, we can just use `Default` for the `None` case

File: crates/storage/provider/src/providers/database/mod.rs [hunk 2]
```diff
@@ -54,8 +56,9 @@ impl<DB> ProviderFactory<DB> {
         db: DB,
         chain_spec: Arc<ChainSpec>,
         static_file_provider: StaticFileProvider,
+        prune_modes: Option<PruneModes>,
```
[the remark this answers is not in the document]
Comment (Rjected) [reply crates/storage/provider/src/providers/database/mod.rs]: removed

File: bin/reth/src/commands/debug_cmd/build_block.rs [hunk 3]
```diff
@@ -122,7 +123,11 @@ impl Command {
         // configure blockchain tree
         let tree_externals =
             TreeExternals::new(provider_factory.clone(), Arc::clone(&consensus), executor);
-        let tree = BlockchainTree::new(tree_externals, BlockchainTreeConfig::default(), None)?;
+        let tree = BlockchainTree::new(
+            tree_externals,
+            BlockchainTreeConfig::default(),
+            PruneModes::default(),
```
Comment (mattsse) [inline bin/reth/src/commands/debug_cmd/build_block.rs:129]:
```suggestion
            PruneModes::none(),
```

File: bin/reth/src/commands/debug_cmd/replay_engine.rs [hunk 4]
```diff
@@ -85,7 +85,11 @@ impl Command {
         // Configure blockchain tree
         let tree_externals =
             TreeExternals::new(provider_factory.clone(), Arc::clone(&consensus), executor);
-        let tree = BlockchainTree::new(tree_externals, BlockchainTreeConfig::default(), None)?;
+        let tree = BlockchainTree::new(
+            tree_externals,
+            BlockchainTreeConfig::default(),
+            PruneModes::default(),
```
Comment (mattsse) [inline bin/reth/src/commands/debug_cmd/replay_engine.rs:91]:
```suggestion
            PruneModes::none(),
```

File: crates/blockchain-tree/src/blockchain_tree.rs [hunk 5]
```diff
@@ -1647,7 +1644,7 @@ mod tests {
         let mut tree = BlockchainTree::new(
             TreeExternals::new(provider_factory, consensus, executor_provider),
             BlockchainTreeConfig::default(),
-            None,
+            PruneModes::default(),
```
Comment (mattsse) [inline crates/blockchain-tree/src/blockchain_tree.rs:1647]:
```suggestion
            PruneModes::none(),
```

**Review by mattsse — APPROVED** [review]
pedantic nit:

let's use `Prunemodes::none` which is just default, but makes it obvious that this is none

**Pull request merged**